# Hoja de Trabajo \# 4

---


por Josué Obregón <br>
DS6011 - Feature Engineering <br>
UVG Masters - Escuela de Negocios<br>


## Objetivos

El objetivo de esta hoja de trabajo  es presentar al estudiante diferentes técnicas de codificación de variables categóricas. Éstas técnicas incluyen codificadores clásicos, de contraste y codificadores supervisados o bayesianos.

También se busca que el estudiante practique la utilización de éstas técnicas con las librerías disponibles en el lenguaje Python.


## Importación de librerías y carga de los datos a varios pandas [DataFrames](https://pandas.pydata.org/pandas-docs/version/1.1.5/reference/frame.html)




In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

In [4]:
!mkdir data

In [5]:
import gdown

urls = ['https://drive.google.com/uc?export=download&id=16AGQw1nM9NYILv2aSZaSNSn9jBPByWPq', # okc_train  https://drive.google.com/file/d/16AGQw1nM9NYILv2aSZaSNSn9jBPByWPq/view?usp=sharing
        ]
outputs = ['okc_train.csv']
for url,output in zip(urls,outputs):
  gdown.download(url, f'data/{output}', quiet=False)

Downloading...
From: https://drive.google.com/uc?export=download&id=16AGQw1nM9NYILv2aSZaSNSn9jBPByWPq
To: /content/data/okc_train.csv
12.3MB [00:00, 108MB/s]


In [8]:
df = pd.read_csv('data/okc_train.csv',index_col=0)

In [9]:
df.head()

age               diet      drinks  ... essay_length  Class  profile
1    22  strictly_anything    socially  ...     3.420945  other        1
4    23         vegetarian    socially  ...     2.671173  other        3
6    29    mostly_anything    socially  ...     3.391112   stem        5
8    31    mostly_anything    socially  ...     3.090611  other        6
10   37    mostly_anything  not_at_all  ...     3.382377  other        7

[5 rows x 38 columns]

In [10]:
df['drinks'].value_counts()

socially          27521
rarely             3981
often              3413
not_at_all         2157
drinks_missing     1208
very_often          319
desperately         210
Name: drinks, dtype: int64

In [ ]:
df['diet'].value_counts()

diet_missing           14698
mostly_anything        11430
anything                4096
strictly_anything       3440
mostly_vegetarian       2313
mostly_other             700
strictly_vegetarian      588
vegetarian               448
strictly_other           304
mostly_vegan             220
other                    217
strictly_vegan           142
vegan                     82
mostly_kosher             59
mostly_halal              39
strictly_kosher           11
strictly_halal            10
kosher                     9
halal                      3
Name: diet, dtype: int64

Pequeño dataset de prueba para algunas explicaciones y para las figuras en las diapositivas


In [11]:
df_test = pd.DataFrame({
... 'City': ['SF', 'SF', 'SF', 'NYC', 'NYC', 'NYC',
... 'Seattle', 'Seattle', 'Seattle'],
... 'Rent': [3999, 4000, 4001, 3499, 3500, 3501, 2499, 2500, 2501]
... })

In [12]:
df_test

City  Rent
0       SF  3999
1       SF  4000
2       SF  4001
3      NYC  3499
4      NYC  3500
5      NYC  3501
6  Seattle  2499
7  Seattle  2500
8  Seattle  2501

Para esta hoja de trabajo vamos a utilizar la librería [category_encoders](http://contrib.scikit-learn.org/category_encoders/index.html), la cual es compatible con scikit-learn.

In [6]:
!pip install category_encoders

     |████████████████████████████████| 81kB 4.7MB/s 


In [13]:
col = 'drinks'

# Unsupervised Encoders

## Classic Encoders

### One-Hot Encoder

In [ ]:
from category_encoders import OneHotEncoder

In [ ]:
one_hot_encoder = OneHotEncoder(use_cat_names=True)

In [ ]:
one_hot_encoder.fit_transform(df[[col]])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks_socially  ...  drinks_desperately
1                    1  ...                   0
4                    1  ...                   0
6                    1  ...                   0
8                    1  ...                   0
10                   0  ...                   0
...                ...  ...                 ...
59940                1  ...                   0
59943                0  ...                   0
59944                0  ...                   0
59945                1  ...                   0
59946                1  ...                   0

[38809 rows x 7 columns]

In [ ]:
one_hot_encoder.mapping[0]['mapping']

drinks_socially  drinks_not_at_all  ...  drinks_very_often  drinks_desperately
 1                1                  0  ...                  0                   0
 2                0                  1  ...                  0                   0
 3                0                  0  ...                  0                   0
 4                0                  0  ...                  0                   0
 5                0                  0  ...                  0                   0
 6                0                  0  ...                  1                   0
 7                0                  0  ...                  0                   1
-1                0                  0  ...                  0                   0
-2                0                  0  ...                  0                   0

[9 rows x 7 columns]

### Binary Encoder

In [ ]:
from category_encoders import BinaryEncoder 

In [ ]:
binary_encoder = BinaryEncoder()

In [ ]:
binary_encoder.fit_transform(df[[col]])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks_0  drinks_1  drinks_2  drinks_3
1             0         0         0         1
4             0         0         0         1
6             0         0         0         1
8             0         0         0         1
10            0         0         1         0
...         ...       ...       ...       ...
59940         0         0         0         1
59943         0         1         0         0
59944         0         0         1         0
59945         0         0         0         1
59946         0         0         0         1

[38809 rows x 4 columns]

In [ ]:
binary_encoder.base_n_encoder.mapping[0]['mapping']

drinks_0  drinks_1  drinks_2  drinks_3
 1         0         0         0         1
 2         0         0         1         0
 3         0         0         1         1
 4         0         1         0         0
 5         0         1         0         1
 6         0         1         1         0
 7         0         1         1         1
-1         0         0         0         0
-2         0         0         0         0

### Frequency or Count Encoder

In [ ]:
from category_encoders import CountEncoder

In [ ]:
count_encoder = CountEncoder(normalize=True)

In [ ]:
count_encoder.fit_transform(df[[col]])

drinks
1      0.709140
4      0.709140
6      0.709140
8      0.709140
10     0.055580
...         ...
59940  0.709140
59943  0.087944
59944  0.055580
59945  0.709140
59946  0.709140

[38809 rows x 1 columns]

In [ ]:
count_encoder = CountEncoder(combine_min_nan_groups=True, min_group_size=500, min_group_name='otros', normalize=True)

In [ ]:
 count_encoder.fit_transform(df[[col]])

drinks
1      0.709140
4      0.709140
6      0.709140
8      0.709140
10     0.055580
...         ...
59940  0.709140
59943  0.087944
59944  0.055580
59945  0.709140
59946  0.709140

[38809 rows x 1 columns]

Intentemos utilizando las funciones de combinacion de grupos

### Ordinal Encoder

In [ ]:
from category_encoders import OrdinalEncoder

In [ ]:
ordinal_encoders = OrdinalEncoder()

In [ ]:
ordinal_encoders.fit_transform(df[[col]])

drinks
1           1
4           1
6           1
8           1
10          2
...       ...
59940       1
59943       4
59944       2
59945       1
59946       1

[38809 rows x 1 columns]

In [ ]:
ordinal_encoders.mapping[0]['mapping']

socially          1
not_at_all        2
rarely            3
often             4
drinks_missing    5
very_often        6
desperately       7
NaN              -2
dtype: int64

###Feature Hashing

In [ ]:
from sklearn.feature_extraction import FeatureHasher 

In [ ]:
col = 'where_town'

In [ ]:
describe()

NameError: ignored

In [ ]:
df[col].head()

1     south_san_francisco
4                berkeley
6           san_francisco
8           san_francisco
10              san_mateo
Name: where_town, dtype: object

In [ ]:
hash_enc = FeatureHasher(n_features=8, input_type='string', alternate_sign=True)

In [ ]:
hashed_features = hash_enc.fit_transform(df[col])

In [ ]:
hashed_features.toarray()

array([[ 1., -1.,  2., ..., -2.,  0., -3.],
       [ 0.,  1.,  0., ..., -1.,  1.,  2.],
       [ 0., -1.,  2., ..., -1.,  0., -2.],
       ...,
       [ 1., -1.,  2., ..., -2.,  0., -3.],
       [ 0., -1.,  2., ..., -1.,  0., -2.],
       [ 0., -1.,  2., ..., -1.,  0., -2.]])

In [ ]:
np_hashed = np.array(hashed_features.todense())

In [ ]:
np.unique(np_hashed, axis=0)

array([[ 0., -2.,  2.,  1.,  0., -2., -1.,  0.],
       [ 0., -1.,  0.,  4.,  0., -1.,  1.,  1.],
       [ 0., -1.,  1.,  0., -1., -3.,  0.,  1.],
       [ 0., -1.,  1.,  1.,  1., -1.,  2.,  0.],
       [ 0., -1.,  2.,  3., -2., -1.,  0., -2.],
       [ 0.,  0.,  0.,  0.,  1., -1.,  0.,  1.],
       [ 0.,  0.,  0.,  1.,  0., -1., -1.,  1.],
       [ 0.,  0.,  0.,  2.,  0., -2.,  0.,  0.],
       [ 0.,  0.,  1., -1.,  0., -1.,  0.,  1.],
       [ 0.,  0.,  1.,  0., -1.,  0.,  0.,  0.],
       [ 0.,  0.,  1.,  0.,  2.,  0.,  0.,  0.],
       [ 0.,  0.,  1.,  0.,  5., -1.,  1.,  1.],
       [ 0.,  0.,  1.,  3., -1.,  0.,  0.,  0.],
       [ 0.,  0.,  2., -1.,  1.,  0.,  1.,  2.],
       [ 0.,  0.,  2.,  0.,  0., -1.,  1.,  0.],
       [ 0.,  0.,  2.,  0.,  2., -2., -1.,  0.],
       [ 0.,  0.,  2.,  0.,  2.,  0.,  1.,  0.],
       [ 0.,  0.,  2.,  1.,  0., -1.,  1.,  0.],
       [ 0.,  0.,  2.,  1.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  2.,  2.,  0., -1., -1., -1.],
       [ 0.,  0.,  2

In [ ]:
np.unique()

In [ ]:
from sys import getsizeof

In [ ]:
print(getsizeof(df[col]))
print(getsizeof(hashed_features))

2944494
64


## Contrast Encoders

### Diferencia entre One-hot Encoding y Dummy Coding

In [ ]:
one_hod_df = pd.get_dummies(df_test, prefix=['city'])

In [ ]:
one_hod_df

Rent  city_NYC  city_SF  city_Seattle
0  3999         0        1             0
1  4000         0        1             0
2  4001         0        1             0
3  3499         1        0             0
4  3500         1        0             0
5  3501         1        0             0
6  2499         0        0             1
7  2500         0        0             1
8  2501         0        0             1

In [ ]:
one_hod_df[['city_NYC', 'city_SF', 'city_Seattle']].drop_duplicates()

city_NYC  city_SF  city_Seattle
0         0        1             0
3         1        0             0
6         0        0             1

In [ ]:
from sklearn import linear_model

In [ ]:
model = linear_model.LinearRegression()
model.fit(one_hod_df[['city_NYC', 'city_SF', 'city_Seattle']], one_hod_df['Rent'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
print(model.coef_)
print(model.intercept_)

[ 166.66666667  666.66666667 -833.33333333]
3333.3333333333335


In [ ]:
print(df_test['Rent'].mean())
print(df_test.groupby('City').mean())

3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


Con la codificación one-hot, el término de intersección (intercepto) representa la media global de la variable objetivo: 'Renta', y cada uno de los coeficientes lineales representa que tanto difiere la renta media de esa ciudad respecto a la media global.

Ahora entrenemos el mismo modelo con dummy coding

Con codificación dummy, el coeficiente de sesgo (intercepto) representa el valor medio de la variable $y$ para la categoría de referencia, que en el ejemplo es la ciudad NYC. El coeficiente para la *i*-ésima característica es igual a la diferencia entre el valor medio de la respuesta para la i-ésima categoría y la media de la categoría de referencia.

In [ ]:
model.fit(dummy_df[['city_SF', 'city_Seattle']], dummy_df['Rent'])
print(model.coef_)
print(model.intercept_)

NameError: ignored

### Dummy Coding


### Sum (or Deviation) Coding

Utilizando el ejemplo de prueba

La codificación de efectos (sum coding) es muy similar a la codificación dummy, pero da como resultado modelos de regresión lineal que son aún más simples de interpretar.

En el ejemplo, vemos que el término de intersección representa la media global del la variable respuesta, y los coeficientes individuales indican cuánto las medias de las categorías individuales difieren de la media global. (Esto se llama el efecto principal de la categoría o nivel, de ahí el nombre "codificación de efectos").

La codificación One-hot, encontró la misma intersección y coeficientes, pero en ese caso hay coeficientes lineales para cada ciudad. En la codificación de efecto, ningun feature representa la categoría de referencia, ***por lo que el efecto de la categoría de referencia debe calcularse por separado como la suma negativa de los coeficientes de todas las demás categorías.***

Restulados de la regresión lineal con One-hot encoding

Coefficients: $[ 166.66666667,   666.66666667,  -833.33333333]$

Intercept: $3333.3333333333335$

### Backward Difference Coding

In [36]:
from category_encoders import BackwardDifferenceEncoder

In [42]:
back_diff_enc = BackwardDifferenceEncoder()
back_diff_df = back_diff_enc.fit_transform(df_test['City'])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



In [39]:
back_diff_enc.mapping[0]['mapping']

City_0    City_1
 1 -0.666667 -0.333333
 2  0.333333 -0.333333
 3  0.333333  0.666667
-1  0.000000  0.000000
-2  0.000000  0.000000

Utilizando el ejemplo de prueba

In [40]:
from sklearn import linear_model

In [43]:
model = linear_model.LinearRegression()
model.fit(back_diff_df[['City_0', 'City_1']], df_test['Rent'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [44]:
print(model.coef_)
print(model.intercept_)

[ -500. -1000.]
3333.3333333333335


In [45]:
print(df_test['Rent'].mean())
print(df_test.groupby('City').mean())

3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


La codificación de diferencia en reversa es útil para codificar variables ordinales.

En el ejemplo, vemos que el término de intersección representa la media global de la variable respuesta, y los coeficientes individuales indican cuánto las medias de las categorías individuales difieren de la media de la categoría inmediatamente anterior.

### Helmert Coding

In [46]:
from category_encoders import HelmertEncoder

In [47]:
hel_enc = HelmertEncoder()
hel_df = hel_enc.fit_transform(df_test['City'], df_test['Rent'])


/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



In [48]:
hel_df

intercept  City_0  City_1
0          1    -1.0    -1.0
1          1    -1.0    -1.0
2          1    -1.0    -1.0
3          1     1.0    -1.0
4          1     1.0    -1.0
5          1     1.0    -1.0
6          1     0.0     2.0
7          1     0.0     2.0
8          1     0.0     2.0

In [49]:
hel_enc.mapping[0]['mapping']

City_0  City_1
 1    -1.0    -1.0
 2     1.0    -1.0
 3     0.0     2.0
-1     0.0     0.0
-2     0.0     0.0

Utilizando el ejemplo de prueba

In [50]:
model = linear_model.LinearRegression()
model.fit(hel_df[['City_0', 'City_1']], df_test['Rent'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [51]:
print(model.coef_)
print(model.intercept_)

[-250.         -416.66666667]
3333.3333333333335


In [52]:
print(df_test['Rent'].mean())
print(df_test.groupby('City').mean())

3333.3333333333335
         Rent
City         
NYC      3500
SF       4000
Seattle  2500


#Supervised Encoders

In [16]:
from sklearn.preprocessing import LabelEncoder

In [17]:
label_enc = LabelEncoder()

df['Class_enc'] = label_enc.fit_transform(df['Class'])

In [18]:
df.head()

age               diet      drinks  ...  Class profile  Class_enc
1    22  strictly_anything    socially  ...  other       1          0
4    23         vegetarian    socially  ...  other       3          0
6    29    mostly_anything    socially  ...   stem       5          1
8    31    mostly_anything    socially  ...  other       6          0
10   37    mostly_anything  not_at_all  ...  other       7          0

[5 rows x 39 columns]

## Target Encoder

In [26]:
from category_encoders import TargetEncoder

In [27]:
cat_enc = TargetEncoder()
cat_enc.fit_transform(df[col_cat], df['Class_enc'])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      0.189201
4      0.189201
6      0.189201
8      0.189201
10     0.147891
...         ...
59940  0.189201
59943  0.195429
59944  0.147891
59945  0.189201
59946  0.189201

[38809 rows x 1 columns]

In [29]:
cat_enc.mapping

{'drinks': drinks
  1    0.189201
  2    0.147891
  3    0.170058
  4    0.195429
  5    0.179636
  6    0.137931
  7    0.171429
 -1    0.184674
 -2    0.184674
 dtype: float64}

## Leave-one-out Encoder

In [30]:
from category_encoders import LeaveOneOutEncoder

In [31]:
loo_enc = LeaveOneOutEncoder()
loo_enc.fit_transform(df[col_cat], df['Class_enc'])

drinks
1      0.189208
4      0.189208
6      0.189172
8      0.189208
10     0.147959
...         ...
59940  0.189208
59943  0.195487
59944  0.147959
59945  0.189208
59946  0.189208

[38809 rows x 1 columns]

In [32]:
loo_enc.mapping

{'drinks':                    sum  count
 desperately       36.0    210
 drinks_missing   217.0   1208
 not_at_all       319.0   2157
 often            667.0   3413
 rarely           677.0   3981
 socially        5207.0  27521
 very_often        44.0    319}

## M-estimate Encoder

In [33]:
from category_encoders import MEstimateEncoder

In [34]:
mes_enc = MEstimateEncoder()
mes_enc.fit_transform(df[col_cat], df['Class_enc'])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      0.189201
4      0.189201
6      0.189201
8      0.189201
10     0.147908
...         ...
59940  0.189201
59943  0.195426
59944  0.147908
59945  0.189201
59946  0.189201

[38809 rows x 1 columns]

In [35]:
mes_enc.mapping

{'drinks': drinks
  1    0.189201
  2    0.147908
  3    0.170061
  4    0.195426
  5    0.179640
  6    0.138077
  7    0.171491
 -1    0.184674
 -2    0.184674
 dtype: float64}

## Weight of Evidence Encoder

In [23]:
from category_encoders import WOEEncoder
col_cat = 'drinks'

In [24]:
woe_enc = WOEEncoder()
woe_enc.fit_transform(df[col_cat], df['Class_enc'])

/usr/local/lib/python3.7/dist-packages/category_encoders/utils.py:21: FutureWarning:

is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead



drinks
1      0.029719
4      0.029719
6      0.029719
8      0.029719
10    -0.263874
...         ...
59940  0.029719
59943  0.070806
59944 -0.263874
59945  0.029719
59946  0.029719

[38809 rows x 1 columns]

In [25]:
woe_enc.mapping

{'drinks': drinks
  1    0.029719
  2   -0.263874
  3   -0.099262
  4    0.070806
  5   -0.030446
  6   -0.328956
  7   -0.069086
 -1    0.000000
 -2    0.000000
 dtype: float64}